### PR TITLE
[FLINK-35479][e2e] Add end-to-end test for materialized table

### PR DESCRIPTION
**Pull request by hackergin**
## What is the purpose of the change

*Add e2e test case for materialized table. * 

## Brief change log

  - *Add e2e test for materialized table*

## Verifying this change
This is a e2e test, and no need to verify.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java [hunk 1]
```diff
@@ -126,6 +135,339 @@ public void testSqlClientExecuteStatement() throws Exception {
...
+    public void testMaterializedTableInFullMode() throws Exception {
+        Duration fullModeWaitTime = Duration.ofMinutes(5);
+        Duration fullModeWaitPause = Duration.ofSeconds(10);
+
+        // init session
+        try (GatewayController gateway = flinkResource.startSqlGateway();
+                ClusterController ignore = flinkResource.startCluster(4)) {
+
+            File catalogStoreFolder = FOLDER.newFolder("catalog-store");
+            Map<String, String> sessionProperties = new HashMap<>();
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:273]: Do we can extract the following code snippet to a utils method?
```
            Map<String, String> sessionProperties = new HashMap<>();
            sessionProperties.put("table.catalog-store.kind", "file");
            sessionProperties.put(
                    "table.catalog-store.file.path", catalogStoreFolder.getAbsolutePath());
            sessionProperties.put("workflow-scheduler.type", "embedded");

            FlinkDistribution.TestSqlGatewayRestClient gatewayRestClient =
                    new FlinkDistribution.TestSqlGatewayRestClient(
                            InetAddress.getByName("localhost").getHostAddress(),
                            restPort.getPort(),
                            SqlGatewayRestAPIVersion.getDefaultVersion().toString().toLowerCase(),
                            sessionProperties);
```

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java [hunk 2]
```diff
@@ -126,6 +135,339 @@ public void testSqlClientExecuteStatement() throws Exception {
...
+            sessionProperties.put("table.catalog-store.kind", "file");
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:274]: The should be configured in Flink config.yaml file? The refresh batch job need to depend on file catalog store to get find materialized table when scheduler open a new session to trigger refresh operation.

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java [hunk 3]
```diff
@@ -126,6 +135,339 @@ public void testSqlClientExecuteStatement() throws Exception {
...
+                                    + " ) with (\n"
+                                    + "    'connector' = 'filesystem',\n"
+                                    + "    'path' = '%s',\n"
+                                    + "    'format' = 'json'\n"
+                                    + " )",
+                            filesystemSourceFolder.getAbsolutePath()));
+
+            gatewayRestClient.executeStatementWithResult(
+                    " CREATE MATERIALIZED TABLE my_materialized_table_in_full_mode\n"
+                            + "                        PARTITIONED BY (ds)\n"
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:316]: ditto

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java [hunk 4]
```diff
@@ -126,6 +135,339 @@ public void testSqlClientExecuteStatement() throws Exception {
...
+            File filesystemSourceFolder = FOLDER.newFolder("filesystem-source");
+
+            gatewayRestClient.executeStatementWithResult(
+                    String.format(
+                            "CREATE TABLE batch_source (\n"
+                                    + "    `timestamp` TIMESTAMP(3),\n"
+                                    + "    `user` VARCHAR,\n"
+                                    + "    `type` VARCHAR\n"
+                                    + " ) with (\n"
+                                    + "    'connector' = 'filesystem',\n"
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:308]: ditto

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java [hunk 5]
```diff
@@ -126,6 +135,339 @@ public void testSqlClientExecuteStatement() throws Exception {
...
+
+            // verify the materialized table should auto refresh the today partition or tomorrow
+            // partition
+            CommonTestUtils.waitUtil(
+                    () -> {
+                        try {
+                            List<RowData> result =
+                                    gatewayRestClient.executeStatementWithResult(
+                                            "select * from my_materialized_table_in_full_mode order by ds");
+                            System.out.println(result);
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:376]: Debug code?

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java [hunk 6]
```diff
@@ -126,6 +135,339 @@ public void testSqlClientExecuteStatement() throws Exception {
...
+                            filesystemSourceFolder.getAbsolutePath()));
+
+            gatewayRestClient.executeStatementWithResult(
+                    " CREATE MATERIALIZED TABLE my_materialized_table_in_full_mode\n"
+                            + "                        PARTITIONED BY (ds)\n"
+                            + "                        WITH (\n"
+                            + "                            'partition.fields.ds.date-formatter' = 'yyyy-MM-dd',\n"
+                            + "                            'format' = 'json'\n"
+                            + "                        )\n"
+                            + "                        FRESHNESS = INTERVAL '30' SECOND\n"
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:321]: Set `1` MINUTE interval for test stability?

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java [hunk 7]
```diff
@@ -126,6 +135,339 @@ public void testSqlClientExecuteStatement() throws Exception {
...
+
+            // suspend the materialized table
+            gatewayRestClient.executeStatementWithResult(
+                    "ALTER MATERIALIZED TABLE my_materialized_table_in_full_mode SUSPEND");
+
+            // resume the materialized table
+            gatewayRestClient.executeStatementWithResult(
+                    "ALTER MATERIALIZED TABLE my_materialized_table_in_full_mode RESUME");
+
+            // insert more data to the batch_source table
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:397]: ditto

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java [hunk 8]
```diff
@@ -126,6 +135,339 @@ public void testSqlClientExecuteStatement() throws Exception {
...
+                            + "                                    DATE_FORMAT(`timestamp`, 'yyyy-MM-dd') AS ds,\n"
+                            + "                                    user,\n"
+                            + "                                    type\n"
+                            + "                                FROM batch_source\n"
+                            + "                            ) GROUP BY ds");
+
+            long systemTime = System.currentTimeMillis();
+            String todayTimestamp =
+                    formatTimestampMillis(systemTime, "yyyy-MM-dd HH:mm:ss", TimeZone.getDefault());
+            String yesterdayTimestamp =
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:337]: What is the function of yesterday?

**Review by lsyldliu — COMMENTED** [review]
@hackergin Thanks for your contribution, I left some comments.

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java (hunk 2 above)
Comment (hackergin) [reply flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java]: Currently, we have saved the initConfig of the session to the workflow info, which can be used by the scheduler for initialization during refresh.

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java (hunk 8 above)
Comment (hackergin) [reply flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:277]: Send data to yerterday for verifying that the refresh operation will not affect historical partitions. I have added relevant verification conditions.

File: flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java [hunk 9]
```diff
@@ -126,6 +134,299 @@ public void testSqlClientExecuteStatement() throws Exception {
...
+                            "yyyy-MM-dd HH:mm:ss",
+                            TimeZone.getDefault());
+            String todayDateStr = todayTimestamp.substring(0, 10);
+            String yesterdayDateStr = yesterdayTimestamp.substring(0, 10);
+            String tomorrowDateStr = tomorrowTimestamp.substring(0, 10);
+
+            // Both send date to current date, yesterday and tomorrow
+            gatewayRestClient.executeStatementWithResult(
+                    String.format(
+                            "INSERT INTO batch_source select TO_TIMESTAMP('%s'), 'Alice', 'INFO'",
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-sql-gateway-test/src/test/java/org/apache/flink/table/gateway/SqlGatewayE2ECase.java:293]: We insert these three records in one query via values?

**Review by lsyldliu — APPROVED** [review]
LGTM

**Pull request merged**